lib/benchmark/bench.go

```diff
@@ -59,175 +57,169 @@ type Result struct {
-			result.Histogram.RecordValue(int64(measure.End.Sub(measure.Start) / time.Millisecond))
+
+		case measure := <-resultC:
 			results = append(results, measure)
 			if timeElapsed && len(results) >= c.MinimumMeasurements {
-				go cancelWorkers()
+				go cancel()
 			}
 			if measure.Error != nil {
 				result.RequestsFailed++
 				result.LastError = measure.Error
 			}
+			result.ResponseHistogram.RecordValue(int64(measure.End.Sub(measure.ResponseStart) / time.Millisecond))
+			result.ServiceHistogram.RecordValue(int64(measure.End.Sub(measure.ServiceStart) / time.Millisecond))
 			result.RequestsOriginated++
-		case <-workerCtx.Done():
+		case <-ctx.Done():
 			result.Duration = time.Since(start)
 			return result, nil
 		case <-statusTicker.C:
-			log.Printf("working... observations: %d", len(results))
+			log.Printf("working... current observation count: %d", len(results))
 		}
+
 	}
 }
 
-func produceMeasures(ctx context.Context, rate int, c chan<- benchMeasure) {
-	interval := time.Duration(float64(1) / float64(rate) * float64(time.Second))
-	ticker := time.NewTicker(interval)
-	defer ticker.Stop()
+type benchMeasure struct {
+	ResponseStart time.Time
+	ServiceStart  time.Time
+	End           time.Time
+	Error         error
+	ctx           context.Context
+	client        *client.TeleportClient
+	command       []string
+	interactive   bool
+}
+
+func run(ctx context.Context, request <-chan *benchMeasure, done chan<- *benchMeasure) {
+	defer func() {
+		if r := recover(); r != nil {
+			logrus.Warningf("recover from panic: %v", r)
+		}
+	}()
+
 	for {
 		select {
-		case <-ticker.C:
 
-			measure := benchMeasure{
-				Start: time.Now(),
-			}
-			select {
-			case c <- measure:
-			case <-ctx.Done():
+		case m, ok := <-request:
+			if !ok {
 				return
 			}
+			go worker(ctx, m, done)
 		case <-ctx.Done():
 			return
 		}
+
 	}
-}
 
-type benchMeasure struct {
-	Start           time.Time
-	End             time.Time
-	ThreadCompleted bool
-	ThreadID        int
-	Error           error
 }
 
-type benchmarkThread struct {
-	id          int
-	ctx         context.Context
-	client      *client.TeleportClient
-	command     []string
-	interactive bool
-	receiveC    chan benchMeasure
-	sendC       chan benchMeasure
+func worker(ctx context.Context, m *benchMeasure, send chan<- *benchMeasure) {
+	m.ServiceStart = time.Now()
+	execute(m)
+	m.End = time.Now()
+	select {
+	case send <- m:
+	case <-ctx.Done():
+		return
+	}
+
 }
 
-func (b *benchmarkThread) execute(measure benchMeasure) {
-	if !b.interactive {
+func execute(m *benchMeasure) {
+	if !m.interactive {
 		// do not use parent context that will cancel in flight requests
 		// because we give test some time to gracefully wrap up
 		// the in-flight connections to avoid extra errors
-		measure.Error = b.client.SSH(context.TODO(), nil, false)
-		measure.End = time.Now()
-		b.sendMeasure(measure)
+
+		m.Error = m.client.SSH(context.TODO(), nil, false)
+		m.End = time.Now()
```

Comment:
`m.End` is set outside of `execute` (line 176).  I don't think we need to set it inside of `execute` as well -- it will just be overwritten.